### PR TITLE
Add computational convexity test via dual KKT comparison

### PR DESCRIPTION
**Pull request by jeffreyhorn**
## Summary

- Implements the feature proposed in `FEATURE_computational_convexity_test.md`: two independent MCP solves (cold-start and warm-start) are compared to detect non-convexity
- Two KKT points with different objective values prove the problem is non-convex (zero false positives, theorem-based)
- Three integration points: standalone script, CLI flag, and pipeline flag

## Implementation

- **Core**: `src/diagnostics/convexity_numerical.py` — `ConvexityResult` dataclass, `check_convexity_numerical()`, `check_convexity_from_results()`, `_compare_results()`
- **Phase 1**: `scripts/check_convexity.py` — standalone script (`python scripts/check_convexity.py model.gms`)
- **Phase 2**: `--check-convexity-numerical` CLI flag (requires `-o` and GAMS)
- **Phase 3**: `--check-convexity` pipeline flag in `run_full_test.py` (records results in model database)
- Feature doc moved to `docs/features/completed/`, README updated

## Test plan

- [x] 13 unit tests for `_compare_results` covering all status combinations and tolerance edge cases
- [x] 2 CLI integration tests (`--help` listing, requires `-o` validation)
- [x] `make format && make lint` pass (ruff + black clean; pre-existing mypy errors in test_solve.py unchanged)
- [x] `make test` — 4444 passed, 0 failed

### CODE REVIEW

File: src/diagnostics/convexity_numerical.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+    Args:
+        mcp_cold_path: Path to cold-start MCP file (no pre-solve).
+        mcp_warm_path: Path to warm-start MCP file (with --nlp-presolve).
+        timeout: GAMS solve timeout in seconds.
+
+    Returns:
+        ConvexityResult with the comparison outcome.
+    """
+    from scripts.gamslib.test_solve import solve_mcp
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/diagnostics/convexity_numerical.py]: Fixed in df73823c. Wrapped the `from scripts.gamslib.test_solve import solve_mcp` in a try/except that raises a clear `ImportError` explaining the module is only available from a source checkout.

File: src/diagnostics/convexity_numerical.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    # Both STATUS 1 with objectives available
+    if status_cold == 1 and status_warm == 1 and obj_cold is not None and obj_warm is not None:
+        abs_diff = abs(obj_cold - obj_warm)
+        denom = max(abs(obj_cold), abs(obj_warm), 1.0)
+        rel_diff = abs_diff / denom
+
+        if rel_diff > rel_tol:
+            return ConvexityResult(
+                is_nonconvex=True,
+                obj_cold=obj_cold,
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/diagnostics/convexity_numerical.py]: Fixed in df73823c. Aligned status handling with pipeline conventions: `_OPTIMAL = {1, 2}` and `_INFEASIBLE = {4, 5}`. STATUS 2 (Locally Optimal) is now treated as optimal for objective comparison, and STATUS 4 (Infeasible) is handled alongside STATUS 5.

File: src/diagnostics/convexity_numerical.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            status_cold=status_cold,
+            status_warm=status_warm,
+            abs_diff=None,
+            rel_diff=None,
+            conclusion="Inconclusive: both solves infeasible (STATUS 5)",
+        )
+
+    # One or both failed to solve
+    cold_status_str = f"STATUS {status_cold}" if status_cold else "no solve"
+    warm_status_str = f"STATUS {status_warm}" if status_warm else "no solve"
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/diagnostics/convexity_numerical.py]: Fixed in df73823c. Changed to explicit `is not None` checks in all three locations (convexity_numerical.py, cli.py, check_convexity.py) so STATUS 0 is reported accurately.

File: scripts/gamslib/run_full_test.py [hunk 4]
```diff
@@ -670,6 +670,69 @@ def _determine_stages(args: argparse.Namespace) -> list[str]:
...
+    model["convexity_numerical"] = {
+        "obj_cold": cvx.obj_cold,
+        "obj_warm": cvx.obj_warm,
+        "status_cold": cvx.status_cold,
+        "status_warm": cvx.status_warm,
+        "is_nonconvex": cvx.is_nonconvex,
+        "abs_diff": cvx.abs_diff,
+        "rel_diff": cvx.rel_diff,
+        "conclusion": cvx.conclusion,
+    }
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply scripts/gamslib/run_full_test.py:736]: Fixed in df73823c. Added `convexity_numerical` property to `model_entry` in schema.json referencing a new `convexity_numerical_result` definition with `additionalProperties: false` and typed fields for obj_cold, obj_warm, status_cold, status_warm, is_nonconvex, abs_diff, rel_diff, and conclusion.

File: src/cli.py [hunk 5]
```diff
@@ -506,6 +513,67 @@ def main(
...
+                    cvx_result = _run_check(cold_path, warm_path)
+
+            if not quiet:
+                click.echo("")
+                click.echo("Computational Convexity Check:")
+                _s_cold = (
+                    f"STATUS {cvx_result.status_cold}" if cvx_result.status_cold else "no solve"
+                )
+                _s_warm = (
+                    f"STATUS {cvx_result.status_warm}" if cvx_result.status_warm else "no solve"
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/cli.py]: Fixed in df73823c. Changed to explicit `is not None` checks for both `status_cold` and `status_warm` in the CLI output formatting.

File: tests/unit/diagnostics/test_convexity_numerical.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+    def test_cold_infeasible_warm_optimal(self):
+        """Cold STATUS 5, warm STATUS 1 → likely non-convex."""
+        cold = {"model_status": 5, "objective_value": None}
+        warm = {"model_status": 1, "objective_value": 1075.547}
+        result = _compare_results(cold, warm)
+        assert not result.is_nonconvex  # Not proven, just likely
+        assert "Likely non-convex" in result.conclusion
+
+    def test_cold_optimal_warm_infeasible(self):
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply tests/unit/diagnostics/test_convexity_numerical.py:78]: Fixed in df73823c. Added 4 new unit tests: `test_status_2_locally_optimal_treated_as_optimal`, `test_both_status_2_same_objective`, `test_status_4_infeasible_cold_warm_optimal`, and `test_status_4_and_5_both_infeasible`. All 4448 tests pass.

File: scripts/check_convexity.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        if not args.quiet:
+            print(f"Solving both MCP files with PATH (timeout={args.timeout}s)...")
+            print()
+
+        result = check_convexity_numerical(cold_path, warm_path, timeout=args.timeout)
+
+        if not args.quiet:
+            print("Computational Convexity Check:")
+            cold_status = f"STATUS {result.status_cold}" if result.status_cold else "no solve"
+            warm_status = f"STATUS {result.status_warm}" if result.status_warm else "no solve"
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply scripts/check_convexity.py]: Fixed in df73823c. Changed to explicit `is not None` checks for both status values in the standalone script output.

File: src/cli.py [hunk 8]
```diff
@@ -506,6 +513,71 @@ def main(
...
 
+        # Step 8: Computational convexity check (optional)
+        if check_convexity_numerical:
+            if not output:
+                click.echo(
+                    "Error: --check-convexity-numerical requires -o OUTPUT",
+                    err=True,
+                )
+                sys.exit(1)
+
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/cli.py:527]: Fixed in e962305f. Moved the `-o` requirement check to the top of the `try` block (before any parsing/conversion) using `raise click.UsageError()`. The CLI now fails immediately with a clear error and no MCP output. Updated the test to verify no MCP code is printed to stdout.

File: scripts/gamslib/run_full_test.py [hunk 9]
```diff
@@ -852,6 +915,10 @@ def run_pipeline(
                 mark_cascade_not_tested(model, "solve", stats)
             return
 
+    # Stage 3b: Computational convexity check (optional)
+    if run_solve and getattr(args, "check_convexity", False):
+        _run_convexity_check(model, model_path, mcp_path, result, args, stats)
+
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply scripts/gamslib/run_full_test.py:941]: Fixed in e962305f. Added `cold_result = result` immediately after the initial `run_solve_stage()` call, before the presolve retry block can overwrite `result`. The convexity check now passes `cold_result` (always the true cold-start) instead of `result` (which may be the warm-start retry).

File: src/cli.py [hunk 10]
```diff
@@ -506,6 +517,64 @@ def main(
...
             click.echo("✓ Conversion complete")
 
+        # Step 8: Computational convexity check (optional)
+        if check_convexity_numerical:
+            import tempfile
+
+            from src.diagnostics.convexity_numerical import (
+                check_convexity_numerical as _run_check,
+            )
+
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/cli.py:527]: Fixed in 2b7edadc. Wrapped the `check_convexity_numerical` import in a try/except that catches `ImportError` and surfaces a clear error message explaining the feature requires a source checkout with `scripts/`.

File: scripts/check_convexity.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    # Check GAMS is available
+    if not shutil.which("gams"):
+        gams_paths = [
+            "/Library/Frameworks/GAMS.framework/Versions/53/Resources/gams",
+            "/Library/Frameworks/GAMS.framework/Versions/Current/Resources/gams",
+        ]
+        if not any(Path(p).exists() for p in gams_paths):
+            print("Error: GAMS not found on PATH", file=sys.stderr)
+            return 1
+
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply scripts/check_convexity.py]: Fixed in 2b7edadc. Removed the redundant GAMS pre-check that only whitelisted two macOS paths. The script now relies on `solve_mcp()` for executable resolution, which checks the full set of platform-specific paths (macOS, Linux, Windows) plus PATH.

File: src/diagnostics/convexity_numerical.py [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+
+Exploits a fundamental property of convex optimization: for a convex
+minimization problem, every KKT point is a global minimum.  Two KKT
+solutions with different objective values prove non-convexity.
+
+The test runs two independent MCP solves:
+1. Cold-start: PATH solves from default initialization (duals = 0).
+2. Warm-start: NLP pre-solve provides primal + dual initialization.
+
+If both reach STATUS 1 (Optimal) with different objectives, the problem
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/diagnostics/convexity_numerical.py]: Fixed in 2b7edadc. Updated the module docstring to say "STATUS 1 or 2 (Optimal/Locally Optimal)" to match the actual `_OPTIMAL = {1, 2}` handling in `_compare_results()`.

File: scripts/gamslib/run_full_test.py [hunk 13]
```diff
@@ -670,6 +670,69 @@ def _determine_stages(args: argparse.Namespace) -> list[str]:
...
+    model_path: Path,
+    mcp_path: Path,
+    cold_result: dict[str, Any],
+    args: argparse.Namespace,
+    stats: dict[str, Any],
+) -> None:
+    """Run computational convexity check after a successful cold-start solve.
+
+    Generates a warm-start MCP (with --nlp-presolve), solves it, and
+    compares objective values.  Two distinct KKT points prove non-convexity.
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply scripts/gamslib/run_full_test.py]: Fixed in 2b7edadc. Updated the docstring to say "after the cold-start solve attempt" and clarify that the cold-start result may represent either a successful or failed solve, since `cold_result` is preserved before the presolve retry.

File: scripts/check_convexity.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env python3
+"""Computational convexity test via dual KKT comparison.
+
+Generates two MCP files (cold-start and warm-start), solves both with
+PATH, and compares objective values.  If both reach STATUS 1 with
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply scripts/check_convexity.py]: Fixed in e519a2e9. Updated the script docstring to say "STATUS 1 or 2 (Optimal/Locally Optimal)".

File: src/cli.py [hunk 15]
```diff
@@ -506,6 +517,73 @@ def main(
...
+                warm_code = emit_gams_mcp(
+                    kkt,
+                    model_name=model_name,
+                    add_comments=add_comments,
+                    config=config,
+                    nlp_presolve=True,
+                    source_file=input_file,
+                )
+                warm_path.write_text(warm_code)
+
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/cli.py:544]: Fixed in e519a2e9. Added `encoding="utf-8"` to `warm_path.write_text(warm_code)` to handle non-ASCII characters (∇, λ, π, ≥) in emit_gams_mcp comments.

File: src/cli.py [hunk 16]
```diff
@@ -506,6 +517,73 @@ def main(
...
+                    cold_path_tmp = Path(tmpdir) / "cold_mcp.gms"
+                    cold_code = emit_gams_mcp(
+                        kkt,
+                        model_name=model_name,
+                        add_comments=add_comments,
+                        config=config,
+                        nlp_presolve=False,
+                    )
+                    cold_path_tmp.write_text(cold_code)
+                    cvx_result = _run_check(cold_path_tmp, warm_path)
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/cli.py]: Fixed in e519a2e9. Added `encoding="utf-8"` to `cold_path_tmp.write_text(cold_code)` for the same reason.

File: scripts/check_convexity.py [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+            print(f"  {result.conclusion}")
+
+        return 0
+
+    finally:
+        if not args.keep_files:
+            import shutil as sh
+
+            sh.rmtree(tmpdir, ignore_errors=True)
+        else:
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply scripts/check_convexity.py]: Fixed in e519a2e9. Changed to `elif not args.quiet:` so the "Generated files kept in" message is suppressed when `--quiet` is set.

File: src/cli.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/cli.py:544]: Fixed in e519a2e9. Same fix as comment #3070624253 — `warm_path.write_text(warm_code, encoding="utf-8")`.

File: src/cli.py [hunk 18]
```diff
@@ -506,6 +517,73 @@ def main(
...
+                    "Error: --check-convexity-numerical requires the MCP solve "
+                    "helper (scripts.gamslib.test_solve), which is only available "
+                    "from a source checkout.",
+                    err=True,
+                )
+                sys.exit(1)
+
+            cold_path = Path(output)
+
+            # Generate warm-start MCP if not already using --nlp-presolve
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/cli.py]: Fixed in e519a2e9. Updated the comment to accurately describe the control flow: warm MCP is always generated, and if the main output already used `--nlp-presolve`, a temporary cold-start MCP is generated below for the comparison.

File: src/diagnostics/convexity_numerical.py [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+            return ConvexityResult(
+                is_nonconvex=False,
+                obj_cold=obj_cold,
+                obj_warm=obj_warm,
+                status_cold=status_cold,
+                status_warm=status_warm,
+                abs_diff=abs_diff,
+                rel_diff=rel_diff,
+                conclusion=(
+                    f"Consistent: same KKT point "
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/diagnostics/convexity_numerical.py]: Fixed in e519a2e9. Changed from "same KKT point" to "Consistent objectives" since equal objective values don't guarantee the same KKT point (multiple global minima can share the same objective value).

File: scripts/check_convexity.py [hunk 20]
```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env python3
+"""Computational convexity test via dual KKT comparison.
+
+Generates two MCP files (cold-start and warm-start), solves both with
+PATH, and compares objective values.  If both reach STATUS 1 with
+different objectives, the problem is provably non-convex.
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply scripts/check_convexity.py]: Fixed in e519a2e9. Updated the script docstring to say "STATUS 1 or 2 (Optimal/Locally Optimal)" matching the `_OPTIMAL = {1, 2}` handling.

File: src/cli.py [hunk 21]
```diff
@@ -181,6 +188,10 @@ def main(
         diag_report = create_report(Path(input_file).name)
 
     try:
+        # Validate option combinations early, before any work is done
+        if check_convexity_numerical and not output:
+            raise click.UsageError("--check-convexity-numerical requires -o OUTPUT")
+
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/cli.py:194]: Fixed in c3d551ef. Added `except click.ClickException: raise` before the broad `except Exception` handler so `click.UsageError` propagates properly with the standard Click usage message and exit code 2.

File: README.md [hunk 22]
```diff
@@ -168,6 +168,7 @@ nlp2mcp input.gms -o output.gms --nlp-presolve
 - `--smooth-abs`: Enable smooth abs() approximation via sqrt(x²+ε)
 - `--smooth-abs-epsilon FLOAT`: Epsilon for abs smoothing (default: 1e-6)
 - `--nlp-presolve`: Solve the original NLP first to warm-start MCP dual variables (helps non-convex models converge)
+- `--check-convexity-numerical`: Run computational convexity test (requires `-o` and GAMS; compares cold-start vs warm-start objectives to detect non-convexity)
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply README.md]: Fixed in c3d551ef. Updated the README to note that `--check-convexity-numerical` requires a source checkout in addition to `-o` and GAMS.

File: scripts/check_convexity.py [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+"""Computational convexity test via dual KKT comparison.
+
+Generates two MCP files (cold-start and warm-start), solves both with
+PATH, and compares objective values.  If both reach STATUS 1 or 2
+(Optimal/Locally Optimal) with different objectives, the problem is
+provably non-convex.
+
+Usage:
+    python scripts/check_convexity.py examples/simple_nlp.gms
+    python scripts/check_convexity.py data/gamslib/raw/bearing.gms --timeout 120
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply scripts/check_convexity.py]: Fixed in c3d551ef. Replaced `data/gamslib/raw/bearing.gms` (not version-controlled) with `examples/simple_nlp.gms` (committed) in the usage example.

File: scripts/gamslib/run_full_test.py [hunk 24]
```diff
@@ -670,6 +670,71 @@ def _determine_stages(args: argparse.Namespace) -> list[str]:
...
+        "abs_diff": cvx.abs_diff,
+        "rel_diff": cvx.rel_diff,
+        "conclusion": cvx.conclusion,
+    }
+
+    # Update stats
+    if cvx.is_nonconvex:
+        stats["convexity_nonconvex"] = stats.get("convexity_nonconvex", 0) + 1
+    else:
+        stats["convexity_consistent"] = stats.get("convexity_consistent", 0) + 1
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply scripts/gamslib/run_full_test.py]: Fixed in c3d551ef. Renamed `convexity_consistent` to `convexity_not_proven_nonconvex` to accurately reflect that this counter includes inconclusive, likely non-convex, and unusual outcomes — not just consistent results.

File: src/diagnostics/convexity_numerical.py [hunk 25]
```diff
@@ -0,0 +1,213 @@
...
+    return ConvexityResult(
+        is_nonconvex=False,
+        obj_cold=obj_cold,
+        obj_warm=obj_warm,
+        status_cold=status_cold,
+        status_warm=status_warm,
+        abs_diff=None,
+        rel_diff=None,
+        conclusion=(f"Inconclusive: cold={cold_status_str}, warm={warm_status_str}"),
+    )
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/diagnostics/convexity_numerical.py]: Fixed in 59717aae. The fallback Inconclusive conclusion now includes error details from `cold_result.get('error')` and `warm_result.get('outcome_category')`, e.g. "Inconclusive: cold=no solve, warm=no solve (cold error: GAMS executable not found; warm error: Timeout after 60 seconds)". Added a test verifying error messages propagate into the conclusion.

File: src/diagnostics/convexity_numerical.py [hunk 26]
```diff
@@ -0,0 +1,213 @@
...
+    status_cold = cold_result.get("model_status")
+    status_warm = warm_result.get("model_status")
+    obj_cold = cold_result.get("objective_value")
+    obj_warm = warm_result.get("objective_value")
+
+    cold_optimal = status_cold in _OPTIMAL
+    warm_optimal = status_warm in _OPTIMAL
+    cold_infeasible = status_cold in _INFEASIBLE
+    warm_infeasible = status_warm in _INFEASIBLE
+
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/diagnostics/convexity_numerical.py:137]: Fixed in 59717aae. Added `_solve_succeeded()` guard that checks `status == 'success'`, `solver_status == 1`, and no `error` field before trusting a result as optimal or infeasible. This prevents false "Non-convex (proven)" conclusions from abnormal solver terminations. Added two tests: `test_failed_solve_not_trusted_as_optimal` and `test_solver_status_not_1_not_trusted`.

File: scripts/gamslib/run_full_test.py [hunk 27]
```diff
@@ -670,6 +670,73 @@ def _determine_stages(args: argparse.Namespace) -> list[str]:
...
+        )
+        if retry_translate["status"] != "success":
+            if args.verbose:
+                logger.info("    [CONVEXITY] Warm-start translation failed, skipping")
+            return
+
+    # Solve warm-start MCP
+    solve_func = get_solve_function()
+    warm_result = solve_func(presolve_path, timeout=60)
+
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply scripts/gamslib/run_full_test.py]: Fixed in 6da1649d. Added `existing_warm_result` parameter to `_run_convexity_check()`. When the presolve retry already succeeded, `warm_retry_result` is captured and passed through, skipping the redundant GAMS/PATH call. When no prior warm result exists (cold-start succeeded without retry), the warm-start MCP is solved as before.

File: src/diagnostics/convexity_numerical.py [hunk 28]
```diff
@@ -0,0 +1,240 @@
...
+        """Return True only for results that indicate a successful solve."""
+        solve_status = result.get("status")
+        if isinstance(solve_status, str) and solve_status.lower() != "success":
+            return False
+        solver_status = result.get("solver_status")
+        if solver_status is not None and solver_status != 1:
+            return False
+        if result.get("error"):
+            return False
+        return True
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/diagnostics/convexity_numerical.py]: Fixed in 4334b6fa. Made `_solve_succeeded()` strict: now requires `status == "success"` AND `solver_status == 1` (missing fields return False). Added `test_missing_fields_not_trusted` to verify bare dicts with only model_status/objective_value are not trusted.

File: src/cli.py [hunk 29]
```diff
@@ -506,6 +517,74 @@ def main(
...
+                )
+            except ImportError:
+                click.echo(
+                    "Error: --check-convexity-numerical requires the MCP solve "
+                    "helper (scripts.gamslib.test_solve), which is only available "
+                    "from a source checkout.",
+                    err=True,
+                )
+                sys.exit(1)
+
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/cli.py]: Fixed in 4334b6fa. Moved the ImportError catch from the `from src.diagnostics... import` statement (which always succeeds) to around the `_run_check()` calls where `check_convexity_numerical()` actually imports `scripts.gamslib.test_solve.solve_mcp` and raises ImportError.

File: scripts/gamslib/run_full_test.py [hunk 30]
```diff
@@ -1358,6 +1445,14 @@ def main() -> int:
         help="Run only solve stage (requires translate success)",
     )
 
+    # Features
+    feature_group = parser.add_argument_group("Features")
+    feature_group.add_argument(
+        "--check-convexity",
+        action="store_true",
+        help="Run computational convexity test on successfully solved models",
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply scripts/gamslib/run_full_test.py]: Fixed in 4334b6fa. Updated the help text to: "Run computational convexity test after the cold-start solve attempt, including models with failed cold solves".

File: src/cli.py [hunk 31]
```diff
@@ -506,6 +517,73 @@ def main(
...
+                warm_code = emit_gams_mcp(
+                    kkt,
+                    model_name=model_name,
+                    add_comments=add_comments,
+                    config=config,
+                    nlp_presolve=True,
+                    source_file=input_file,
+                )
+                warm_path.write_text(warm_code, encoding="utf-8")
+
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/cli.py:544]: Fixed in 62ad801b. Added `encoding="utf-8"` to the main `output_path.write_text(gams_code)` call in Step 7, matching the temporary MCP writes in the convexity check block.

File: src/diagnostics/convexity_numerical.py [hunk 32]
```diff
@@ -0,0 +1,242 @@
...
+        Requires both status="success" and solver_status=1 with no error.
+        Missing fields are treated as failure (conservative).
+        """
+        if result.get("status") != "success":
+            return False
+        if result.get("solver_status") != 1:
+            return False
+        if result.get("error"):
+            return False
+        return True
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/diagnostics/convexity_numerical.py]: Fixed in 0c2dcbaa. Split into two trust predicates: `_solve_optimal()` (strict — requires status="success", solver_status=1, no error) for the non-convexity proof path, and `_solver_completed()` (requires solver_status=1 only) for infeasible classification. Infeasible branches are now reachable with real solve_mcp outputs where status="failure" but solver_status=1.

File: tests/unit/diagnostics/test_convexity_numerical.py [hunk 33]
```diff
@@ -0,0 +1,222 @@
...
+    ConvexityResult,
+    _compare_results,
+    check_convexity_from_results,
+)
+
+pytestmark = pytest.mark.unit
+
+
+def _ok(model_status: int, objective_value: float | None = None) -> dict:
+    """Build a successful solve result dict."""
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply tests/unit/diagnostics/test_convexity_numerical.py]: Fixed in 0c2dcbaa. Updated `_ok()` to return `status="failure", solver_status=1` for model_status 4/5, mirroring the actual solve_mcp output shape for infeasible solves. All existing tests pass with the updated fixtures.
